### PR TITLE
Allow ShopRow format for coinmasters

### PR DESCRIPTION
**Pull request by Veracity0**
1) I souped up the "test shoprow" command to print shop rows in new coinmaster format.
For example:
```
> test load test_shop_kiwi.html
> test shoprows
```
prints this:
```
--------------------
Kiwi Kwiki Mart	ROW1477	mini kiwi bikini	mini kiwi (2)
Kiwi Kwiki Mart	ROW1488	mini kiwi icepick	mini kiwi (4)
Kiwi Kwiki Mart	ROW1487	mini kiwi whipping stick	mini kiwi (4)
Kiwi Kwiki Mart	ROW1480	mini kiwi invisible dirigible	mini kiwi (3)
Kiwi Kwiki Mart	ROW1482	mini kiwi silicon tiepin	mini kiwi (3)
Kiwi Kwiki Mart	ROW1484	mini kiwi intoxicating spirits	mini kiwi (3)
Kiwi Kwiki Mart	ROW1478	mini kiwitini	mini kiwi	bottle of vodka
Kiwi Kwiki Mart	ROW1479	mini kiwitini	mini kiwi	bottle of gin
Kiwi Kwiki Mart	ROW1486	mini kiwi antimilitaristic hippy petition	mini kiwi (3)
Kiwi Kwiki Mart	ROW1485	mini kiwi illicit antibiotic	mini kiwi (3)
Kiwi Kwiki Mart	ROW1483	mini kiwi tipi	mini kiwi
Kiwi Kwiki Mart	ROW1489	mini kiwi-flavored aspirin-injecting lipstick	mini kiwi (3)
Kiwi Kwiki Mart	ROW1481	mini kiwi aioli	mini kiwi (5)
Kiwi Kwiki Mart	ROW1490	mini kiwi digitized cookies	mini kiwi (3)
--------------------
```
which would be useful if we wanted to turn that creation method into a coinmaster.
Other examples:

star chart
```
--------------------
Star Chart	ROW141	Richard's star key	star (8)	line (7)	star chart
Star Chart	ROW138	star buckler	star (4)	line (6)	star chart
Star Chart	ROW137	star hat	star (5)	line (3)	star chart
Star Chart	ROW136	star pants	star (7)	line (7)	star chart
Star Chart	ROW140	star starfish	star (6)	line (4)	star chart
Star Chart	ROW145	star spatula	star (5)	line (5)	star chart
Star Chart	ROW144	star boomerang	star (4)	line (5)	star chart
Star Chart	ROW143	star stiletto	star (5)	line (4)	star chart
Star Chart	ROW134	star crossbow	star (5)	line (6)	star chart
Star Chart	ROW135	star staff	star (6)	line (5)	star chart
Star Chart	ROW133	star sword	star (7)	line (4)	star chart
Star Chart	ROW139	star throwing star	star (4)	line (2)	star chart
Star Chart	ROW142	star shirt	star (15)	line (15)	star chart
--------------------
```
grandma
```
--------------------
Grandma Sea Monkee's Window	ROW131	crappy Mer-kin mask	Mer-kin scholar mask
Grandma Sea Monkee's Window	ROW124	crappy Mer-kin mask	aerated diving helmet	pristine fish scale (3)
Grandma Sea Monkee's Window	ROW125	crappy Mer-kin tailpiece	sea chaps	teflon swim fins	pristine fish scale (3)
Grandma Sea Monkee's Window	ROW126	Mer-kin gladiator mask	crappy Mer-kin mask	Mer-kin headguard
Grandma Sea Monkee's Window	ROW127	Mer-kin gladiator tailpiece	crappy Mer-kin tailpiece	Mer-kin thighguard
Grandma Sea Monkee's Window	ROW129	Mer-kin scholar mask	crappy Mer-kin mask	Mer-kin facecowl
Grandma Sea Monkee's Window	ROW130	Mer-kin scholar tailpiece	crappy Mer-kin tailpiece	Mer-kin waistrope
Grandma Sea Monkee's Window	ROW116	sushi doily	sea lace
Grandma Sea Monkee's Window	ROW117	scimitar cozy	sea lace (3)	pristine fish scale
Grandma Sea Monkee's Window	ROW118	fish stick cozy	sea lace (3)	pristine fish scale
Grandma Sea Monkee's Window	ROW119	bazooka cozy	sea lace (3)	pristine fish scale
Grandma Sea Monkee's Window	ROW132	scale-mail underwear	dull fish scale (25)	pristine fish scale
Grandma Sea Monkee's Window	ROW120	sea mantle	pristine fish scale (5)
Grandma Sea Monkee's Window	ROW121	sea shawl	pristine fish scale (5)
Grandma Sea Monkee's Window	ROW122	sea cape	pristine fish scale (5)
Grandma Sea Monkee's Window	ROW123	saltwaterbed	sea leather (15)
--------------------
```
And one might think that the Crackpot Mystic should also work, but it is subtly different: instead of saying "green pixel (10)" with 2 "td" tags it says "10 green pixel", using 3 "td" tags.

I'll need to fix that - after finishing the Crimbo24 stores. I have some html for it, but I need to get some with a yellow pixel in inventory.

Anyway, time to move on supporting those kinds of rows...

### CODE REVIEW

**Comment by Veracity0** [comment]
1) CoinmastersDatabase can now parse ShopRow formatted entries.
2) CoinmasterData allows Coinmasters to use specify that they want thse shopRows
3) Crimbo24Bar, Crimbo24Cafe, and Crimbo24Factory no longer use buyItems and buyPrices
4) I am working on the CoinmastersFrame to work with Coinmasters that have shopRows.

Currently, when I start the CoinmastersFrame, debug logging prints this:
```
Crimbo24 Bar has 4 items for sale
Crimbo24 Bar	ROW1521	Easter grasshopper	Spirit of Easter (15)
Crimbo24 Bar	ROW1522	Irish eggnog	Spirit of St. Patrick's Day (15)
Crimbo24 Bar	ROW1523	canteen of jungle juice	Spirit of Veteran's Day (15)
Crimbo24 Bar	ROW1524	turchucken	Spirit of Thanksgiving (15)
Crimbo24 Cafe has 4 items for sale
Crimbo24 Cafe	ROW1527	chocolate ostrich egg	Spirit of Easter (15)
Crimbo24 Cafe	ROW1528	candied beef and cabbage	Spirit of St. Patrick's Day (15)
Crimbo24 Cafe	ROW1529	candy rations	Spirit of Veteran's Day (15)
Crimbo24 Cafe	ROW1530	double-candied yams	Spirit of Thanksgiving (15)
Crimbo24 Factory has 10 items for sale
Crimbo24 Factory	ROW1538	lucky moai statuette	Spirit of Easter (50)	Spirit of St. Patrick's Day (50)
Crimbo24 Factory	ROW1537	egg gun	Spirit of Easter (200)	Spirit of Veteran's Day (200)
Crimbo24 Factory	ROW1540	lucky army helmet	Spirit of St. Patrick's Day (350)	Spirit of Veteran's Day (75)
Crimbo24 Factory	ROW1541	candy egg deviler	Spirit of Easter (200)	Spirit of Thanksgiving (125)
Crimbo24 Factory	ROW1542	lucky napkin	Spirit of St. Patrick's Day (50)	Spirit of Thanksgiving (25)
Crimbo24 Factory	ROW1543	Thanksgiving ration	Spirit of Veteran's Day (25)	Spirit of Thanksgiving (50)
Crimbo24 Factory	ROW1533	Secrets of the Master Egg Hunters	Spirit of Easter (500)
Crimbo24 Factory	ROW1534	How to Lose Friends and Attract Snakes	Spirit of St. Patrick's Day (500)
Crimbo24 Factory	ROW1535	Covert Ops for Kids	Spirit of Veteran's Day (500)
Crimbo24 Factory	ROW1536	Holiday Multitasking	Spirit of Thanksgiving (500)
```
Which tells me that items 1, 2, and 3 are working correctly.

Unfortunately, the items are not showing in the panel yet.
And, I am sure that once I get them to display, they won't work, since I have not added all the code to CoinmasterData and CoinMasterRequest to handle ShopRows yet.

I'm commiting what I have so far, and will continue working on what remains.
Starting with the CoinmastersFrame at least shoping me the items.

**Comment by Veracity0** [comment]
I've fixed the display issues with ShopRow coinmasters and the CoinmastersFrame.

Upcoming:

- write tests for NPCPurchaseRequest.parseShopInventory()
- make sure I can select items and choose quantities to create
- generate requests that will do that.
- register requests for Coinmasters that use ShopRows
- manipulate inventory to remove spent items after such requests succeed.
Those COULD be a new PR, so I am marking this ready to review

File: src/net/sourceforge/kolmafia/ShopRow.java [hunk 1]
```diff
@@ -317,4 +328,52 @@ public String toString() {
       return KoLConstants.COMMA_FORMAT.format(this.count) + " Meat";
     }
   }
+
+  // Conversion from ShopRow objects to and from data strings.
+  // KoLmafia data file format is tab-separated fields
+  //
+  // Grandma Sea Monkee's Window	ROW124	crappy Mer-kin mask	aerated diving helmet	pristine fish
+  // scale (3)
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/ShopRow.java:336]: I recognise that this is spotless screwing things up but a shorter row might appear nicer?

**Review by midgleyc — APPROVED** [review]

**Comment by Veracity0** [comment]
Huh.

- It shouldn't show rows it already knows
- It should only use the old format if it's an existing shop that uses buyItems/sellItems.

I have a couple other small things that will go in a short followup PR.

Thanks!

**Pull request merged**